### ISSUE TITLE
Implement cross-validation in DeepChem

### ISSUE THREAD

**Issue by SeanXiaoyuSun** [deepchem/deepchem#1545]
**Description:**

Dear all,

I'm trying to make an implementation of cross-validation based on post #736 and #926. A subtle difference here is that I split the training set and test set manually. So I loaded two data sets in my code. My implementation seems to be working. But I got some really really good results (using single train-valid-test splitter gave me an RMSE of 30 but this implementation gave an RMSE of 13). I don't think this result is reasonable since I only changed the splitter. Could anyone have a look of my code and give me any suggestions about what's wrong with my implementation? Thanks!

Here is my code:
```
data_dir = "my data_dir"

train_dataset_file = os.path.join(data_dir, "my training data file")
test_dataset_file = os.path.join(data_dir, "my test data file")

if not os.path.exists(train_dataset_file) or not os.path.exists(test_dataset_file):
    print("data set was not found in the given directory")

flashpoint_tasks = ['flashPoint']

loader = deepchem.data.CSVLoader(
    tasks=flashpoint_tasks, smiles_field="smiles", featurizer = deepchem.feat.ConvMolFeaturizer())
train_dataset = loader.featurize(train_dataset_file, shard_size=8192)
test_dataset = loader.featurize(test_dataset_file, shard_size=8192)
    
# Initialize transformers
transformers = [
    deepchem.trans.NormalizationTransformer(
    transform_y=True, dataset=train_dataset, move_mean=True)
]
for transformer in transformers:
    train_dataset = transformer.transform(train_dataset)

transformers = [
    deepchem.trans.NormalizationTransformer(
    transform_y=True, dataset=test_dataset, move_mean=True)
]
for transformer in transformers:
    test_dataset = transformer.transform(test_dataset)
    
splitter = deepchem.splits.RandomSplitter()
split_datas = splitter.k_fold_split(train_dataset,5)

batch_size = 50
metric = dc.metrics.Metric(dc.metrics.rms_score, np.mean)
validation_scores = list()
train_scores = list()
model = dc.models.GraphConvModel(len(flashpoint_tasks), mode='regression')
for train_set, val_set in split_datas:
    model.fit(train_set, nb_epoch= 50)
    train_score = model.evaluate(train_set, [metric], transformers)
    train_scores.append(train_score)
    valid_score = model.evaluate(val_set, [metric], transformers)
    validation_scores.append(valid_score)   
    
print("cross_test_score: ")
test_score = model.evaluate(test_dataset, [metric], transformers)

```

**Comment by SeanXiaoyuSun** [comment]
Also, I noticed that @rbharath said "Each split should have its own model object. The current design leaks information since weights are shared" in post #926. If we want to conduct cross validation, shouldn't we a single model object train these splits and test this model object by using a test data set? Sorry for the novice question.

**Comment by zealseeker** [comment]
Hi Sean,
In your code, the (single) model would be fitted five times by the five `train_set`, which are the five "4/5 of the whole dataset".
Let's imagine that in the second loop, the `train_set` becomes `[0, 2, 3, 4]` while the val_set is `[1]`. However, the model object has already trained in the first loop, in which the `train_set` was `[0, 1, 2, 3]`. Say, the model has been trained by all the data in the whole data set and the performance will be overestimated. In real cross validation, we want to use the val_set `[1]` to validate the performance of the model trained by `[0, 2, 3, 4]`.
We should never make model trained by validation set.

**Comment by SeanXiaoyuSun** [comment]
Thank you for your explanation @zealseeker ! 
I'm confused about how to conduct cross-validation now. As you explained, when doing CV, we should give every split a model object, which means we don't have learned model left after conducting CV. Then which model object should we use to run the test data set? Should we train a model object by the training set again and run test set on this learned model?

**Comment by zealseeker** [comment]
@SeanXiaoyuSun You are right. Cross-validation is used to evaluate the rubostness of the "final model" trained by the whole training set. However, during cross-validation, the `n` models actually are not trained by the whole training set, but `n-1/n` for each fold.

**Comment by SeanXiaoyuSun** [comment]
Thanks! 
I'm gong to close this question now. Feel free to reopen!
